File: Programmers/disk_controller.py
import heapq


def solution(jobs):
    answer = 0
    time = 0
    jobs_len = len(jobs)

    jobs.sort(key=lambda x: (x[1], x[0]))  # 작업 소요시간 순으로 정렬하고 같으면 도착 시간순으로 정렬.

    while len(jobs) != 0:

        for i in range(len(jobs)):
            if jobs[i][0] <= time:  # 현재 시각 기준 도착해있는 것들 중 가장 소요시간 짧은거(이미 정렬됨) 선택해서 처리.
                answer += (time - jobs[i][0] + jobs[i][1])  # 총 작업시간 계산해서 저장.
                time += jobs[i][1]  # 선택한 작업 반영하여 현재 시각 초기화.
                jobs.remove(jobs[i])
                break
            if i == len(jobs) - 1:  # 현재 시각 기준 도착한 작업이 없으면 1초 대기.
                time += 1

    answer = answer // jobs_len

    return answer

File: Java/src/Programmers/DiskController.java
package Programmers;

import java.util.*;

class Task {
    private int requestTime; // 요청시간
    private int turnaroundTime; // 소요시간

    public Task(int requestTime, int turnaroundTime) {
        this.requestTime = requestTime;
        this.turnaroundTime = turnaroundTime;
    }

    public int getRequestTime() {
        return requestTime;
    }

    public int getTurnaroundTime() {
        return turnaroundTime;
    }
}


public class DiskController {
    public int solution(int[][] jobs) {
        int curTime = 0; // 현재시각
        List<Integer> result = new ArrayList<>();
        List<Task> tasks = new ArrayList<>();

        for (int i = 0; i < jobs.length; i++) {
            tasks.add(new Task(jobs[i][0], jobs[i][1]));
        }

        // 요청시간이 이른순, 소요시간 짧은순으로 정렬.
        tasks.sort(
                (task1, task2) -> {
                    int compareByRequestTime = Integer.compare(task1.getRequestTime(), task2.getRequestTime());

                    if (compareByRequestTime != 0) {
                        return compareByRequestTime;
                    }
                    else {
                        return Integer.compare(task1.getTurnaroundTime(), task2.getTurnaroundTime());
                    }
                }
        );

        Deque<Task> waitQ = new ArrayDeque<>(tasks);

        // 하나의 작업이 실행되는 동안 이미 들어온 작업들은 소요시간 짧은 것 부터 우선적으로 처리.
        PriorityQueue<Task> readyQ = new PriorityQueue<>(
                (task1, task2) -> {
                    int compareByTurnaroundTime = Integer.compare(task1.getTurnaroundTime(), task2.getTurnaroundTime());

                    if (compareByTurnaroundTime != 0) {
                        return compareByTurnaroundTime;
                    }
                    else {
                        return Integer.compare(task1.getRequestTime(), task2.getRequestTime());
                    }
                }
        );

        while (!waitQ.isEmpty() || !readyQ.isEmpty()) {

            // 처리할 요청이 없으면 waitQ에서 하나 가져와서 처리.
            if (readyQ.isEmpty()) {
                curTime = waitQ.element().getRequestTime();
                readyQ.add(waitQ.remove());
            }

            Task task = readyQ.remove();
            result.add(curTime - task.getRequestTime() + task.getTurnaroundTime());
            curTime += task.getTurnaroundTime();

            // 하나의 작업을 처리하는 동안 들어온 요청들을 우선순위 큐에 넣어줌.
            while (!waitQ.isEmpty() && waitQ.element().getRequestTime() <= curTime) {
                readyQ.add(waitQ.remove());
            }
        }

        return (int)result.stream().mapToInt(Integer::intValue).average().orElse(0.0);
    }
}
